tmt/base.py

```diff
@@ -3649,10 +3658,15 @@ def from_spec(cls, spec: _RawLink) -> 'Link':
         # `spec` can be either a string, fmf id, or relation:target mapping with
         # a single key (modulo `note` key, of course).
 
-        # String is simple: if `spec` is a string, it represents a target,
-        # and we use the default relationship.
+        # String is simple: if `spec` is a string, it represents a [relation:]target,
+        # and we use the default relationship if relation is not specified. Note that
+        # target should not include ':', that is, the relation must be provided if
+        # target includes ':'.
         if isinstance(spec, str):
-            return Link(relation=Link.DEFAULT_RELATIONSHIP, target=spec)
+            parts = spec.split(':', maxsplit=1)
```

Comment:
This could fail if url contains a ':' and there is no explicit relation, e.g. IPv6 url without fqdn, not sure if this is something we care to solve?

Reply:
Actually every url that starts with http:// would get relation matched as http, so IMO more precise matching needs to be used.

Reply:
Yes, **more precise matching**needs to be used. Currently  I added Link._relation to help to validate the link relation.

Reply:
The `split` will not be enough, `--link http://foo.redhat.com` would be parsed incorrectly. You may check whether there are *2* `:` characters in the raw link, but then we'd fail to handle `--link verifies:/foo` which is also a valid link.

I really don't see a better way than regular expressions, you can make the relation part & `:` optional, and accept everything else. And it can be constructed from `Links._relations`:

```python
>>> import re
>>> from tmt.base import Links
>>> Links._relations
['verifies', 'verified-by', 'implements', 'implemented-by', 'documents', 'documented-by', 'blocks', 'blocked-by', 'duplicates', 'duplicated-by', 'parent', 'child', 'relates', 'test-script']
>>> pattern = re.compile(rf'(?:(?P<relation>{"|".join(Links._relations)}):)?(?P<target>.+)')
>>> pattern
re.compile('(?:(?P<relation>verifies|verified-by|implements|implemented-by|documents|documented-by|blocks|blocked-by|duplicates|duplicated-by|parent|child|relates|test-script):)?(?P<target>.+)')
>>> pattern.match('/foo').groupdict()
{'relation': None, 'target': '/foo'}
>>> pattern.match('verifies:/foo').groupdict()
{'relation': 'verifies', 'target': '/foo'}
>>> pattern.match('verifies/foo').groupdict()
{'relation': None, 'target': 'verifies/foo'}
>>> pattern.match('verifies:http://foo').groupdict()
{'relation': 'verifies', 'target': 'http://foo'}
>>> pattern.match('http://foo').groupdict()
{'relation': None, 'target': 'http://foo'}
>>> 
```

Reply:
Right, it's even in your test:

```
(tmt) huanli@KVM:tests$ # 02 - one link with ':' + invalid link relation
(tmt) huanli@KVM:tests$ rm -rf foo01
(tmt) huanli@KVM:tests$ tmt test create /tests/foo01 --link https://issues.redhat.com/browse/TT-123
Template (shell or beakerlib): shell
Test directory '/home/huanli/dev/368/foo/tests/foo01' created.

Invalid link relation 'https' (should be verifies, verified-by, implements, implemented-by, documents, documented-by, blocks, blocked-by, duplicates, duplicated-by, parent, child, relates or test-script).
```

`https` was not input as a relation, the default relation should have been used in this case. The user provided a link and `split()` is not smart enough to do the right job, which leads to link not being accepted & a misleading message. We need a bigger boat :)

Reply:
Hi @happz，thank you very much for writing the pattern, which helps a lot as I'm not very familiar with writing regex.

```python
pattern = re.compile(rf'(?:(?P<relation>{"|".join(Links._relations)}):)?(?P<target>.+)')
```

Now the related code block is updated according to your suggestion,  please help to review again, thanks！

Reply:
Nice, thanks. Let's reduce the code a bit :)

```python
            result = pattern.match(spec)
            if result is None:
                raise tmt.utils.SpecificationError(
                    f"Invalid spec '{spec}' (should be [relation:]<target>).")
            relation_target_pair = result.groupdict()
            relation = relation_target_pair.get('relation', None)
            target = cast(str, relation_target_pair.get('target', None))
            if relation is None:
                relation = Link.DEFAULT_RELATIONSHIP
            else:
                relation = cast(_RawLinkRelationName, relation)
            if relation not in Links._relations:
                raise tmt.utils.SpecificationError(
                    f"Invalid link relation '{relation}' (should be "
                    f"{fmf.utils.listed(Links._relations, join='or')}).")
```

* `compile` is not necessary, the pattern is not cached in our code anyway, we can use `re.match()` directly
* `relation = cast(_RawLinkRelationName, relation_target_pair['relation'] or Link.DEFAULT_RELATIONSHIP)` - `'relation'` key will exists, always, but it might be set to `None` if the group did not match anything. We can use that, avoid `get`, and use `or` to include the default relationship. Together with `cast()`, `relation` woudl be ready after a single line.
* `target = relation_target_pair['target']` - here we should probably do `assert relation_target_pair['target']` rather than `get()` - the group should always match something, and it should not be empty, regex clearly states `.+` after colon, therefore there must be at least a character. mypy can't tell, the key is seen as `Optional[str]`, `assert` should narrow the type. `get()` is better for situations where the key may or may not exist - we know it must, we just need to tell that to mypy :)
* and I'd drop the `relation not in` check - thanks to regex, constructed from the list, we can be sure the `relation` comes from the list. Otherwise, the regex wouldn't match.

```python
assert relation_target_pair['target'] is not None

relation = cast(_RawLinkRelationName, relation_target_pair['relation'] or Link.DEFAULT_RELATIONSHIP)
target = relation_target_pair['target']

return Link(relation=relation, target=target)
```

:) Correct me if there's any mistake.